core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/compression/GzipCompressor.java

```diff
@@ -38,22 +38,39 @@ public String compressorType() {
     }
 
     @Override
-    public InputStream compress(InputStream inputStream) {
+    public byte[] compress(byte[] content) {
         try {
-            byte[] content = IoUtils.toByteArray(inputStream);
             ByteArrayOutputStream compressedOutputStream = new ByteArrayOutputStream();
             GZIPOutputStream gzipOutputStream = new GZIPOutputStream(compressedOutputStream);
             gzipOutputStream.write(content);
             gzipOutputStream.close();
+            return compressedOutputStream.toByteArray();
+        } catch (IOException e) {
+            throw SdkClientException.create(e.getMessage(), e);
+        }
+    }
 
-            return new ByteArrayInputStream(compressedOutputStream.toByteArray());
+    @Override
+    public InputStream compress(InputStream inputStream) {
+        try {
+            byte[] content = IoUtils.toByteArray(inputStream);
+            byte[] compressedContent = compress(content);
+            return new ByteArrayInputStream(compressedContent);
         } catch (IOException e) {
             throw SdkClientException.create(e.getMessage(), e);
         }
     }
```

Comment:
We should really favor not loading stuff into memory if we can avoid it. Why not just do return `new GZIPInputStream(inputStream)`?

Reply:
`GZIPInputStream` is actually for uncompressing (naming is confusing), we need to provide an output stream to `GZIPOutputStream` and then call `write(byte[] contentToCompress)`